hbase-common/src/test/java/org/apache/hadoop/hbase/HBaseCommonTestingUtility.java

```diff
@@ -264,4 +265,34 @@ boolean deleteDir(final File dir) {
       boolean failIfTimeout, Predicate<E> predicate) throws E {
     return Waiter.waitFor(this.conf, timeout, interval, failIfTimeout, predicate);
   }
+
+  private static final PortAllocator portAllocator = new PortAllocator();
+
+  public static int randomFreePort() {
+    try {
+      return portAllocator.randomFreePort();
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  static class PortAllocator {
+
+    public PortAllocator() {
+
+    }
+
+    public int randomFreePort() throws IOException {
+      ServerSocket s = new ServerSocket(0);
+      try {
+        s.setReuseAddress(true);
+        int port = s.getLocalPort();
+        return port;
+      } finally {
+        if (null != s) {
+          s.close();
+        }
+      }
+    }
+  }
 }
```

Comment:
This is funny. I did this moving back of this facility in HBASE-24307

Reply:
I don't think this is enough though. Still need to deal w/ BindException happening.

Reply:
Yeah, I'm not too concerned how that test itself is addressed, but I think it's good to cleanup getting the port. There are multiple implementations and the one currently used in this test is week and more complicated vs the other of just using port 0 (and properly using setReuseAddrress=true.

You have to also still catch a bind exception because there is a race - it's better to specify 0 for the socket that will actually be used, not see that a port is free and then try to use it (may not be free a second later).

Your solution is isolated to that test though right? The tests that use these other 2 get port methods have the same race, so it seems better to combine the get free port methods into one and make that robust?

Reply:
I suppose it's hard to make robust universally though. It's a problematic way to get a free ephem port. Just trying to address at these common get free port methods in some better way since they are available and used in a few places.

Reply:
Let me try and summarize that:
I think your patch attempts to solve the issue for this test in a solid way.
With my patch, I'm trying to climb into these getPort util methods and make them more solid - though they cannot necessarily be made as solid.

So I don't think the two things are necessarily exclusive.

Reply:
Lets get rid of all impls except this one that you and I moved back into common testing utils.

We get the port number but it could be taken by another by the time we get to use it so handling bindexception seems needed everywhere. Catching has been generalized a bunch but still a few places where we don't catch the bind exception.